### PR TITLE
[s] You can no longer Summon Cultist to get out of Admin Prison

### LINKED ISSUE #60968: Cult teleport working inside admin prison
## Round ID:
[168225](https://scrubby.melonmesa.com/round/168225)

## Reproduction:

Player was able to escape admin prison via some method of cult teleport. They then rejoined the shift and continued to play as normal.

### PR DESCRIPTION
**Pull request by san7890**
Remember: [s] means either [spam](https://github.com/tgstation/tgstation/issues/62349) (Oranges) or [secret](https://discord.com/channels/326822144233439242/326831214667235328/960687157087965185) (MSO). It never means *security*, although we might think it's used that way, it's just not what it means.

## About The Pull Request

Hey there,

There's an old issue report that detailed how, using the Summon Cultist rune, you could be able to escape the Admin Prison (!!!) from CentCom (and could even go back to the station from CentCom using this in general!). This is very bad! So, let's fix this by turning it into a do_teleport rather than a forcemove.

## Why It's Good For The Game
Fixes #60968

## Changelog

:cl:
admin: If you send a cultist to Admin Prison, their fellow cultists are no longer able to summon them out using the Summon Cultist rune.
/:cl:

### CODE REVIEW

File: code/modules/admin/topic.dm [hunk 1]
```diff
@@ -512,6 +512,7 @@
 			return
 
 		M.forceMove(pick(GLOB.prisonwarp))
+		M.AddComponent(/datum/component/imprisoned)
```
Comment (JohnFulpWillard) [inline code/modules/admin/topic.dm:515]: how would you undo this after?

**Review by JohnFulpWillard — COMMENTED** [review]
make the rune teleport instead of forcemove, that way you can make it respect centcom teleporting like everything else does (including wizard)

File: code/modules/admin/topic.dm (hunk 1 above)
Comment (san7890) [reply code/modules/admin/topic.dm:515]: > how would you undo this after?

See the imprisoned tracker, it does the same thing as that itembound component I did earlier this year. Whenever the move_tracker updates, it callbacks to a proc that verifies if the area the prisoner is currently in is still that same area that we have defined in code (admin prison cells). If the areas mismatch, we delete the component and remove the trait.

**Review by MrMelbert — CHANGES_REQUESTED** [review]
Feels over engineered, just make it so summon cultist uses `do_teleport` (blocked by admin prison noteleport) instead of a cringe forcemove

edit: willard sniped it before me

**Comment by san7890** [comment]
I'm probably going to leave the framework already introduced in this PR for two reasons:

A) Having that specific check for TRAIT_NO_TELEPORT in the Summon Cultist chain allows us to give a specific reason in game.log as to the failure of summoning that cultist, in case that becomes valuable later.

B) I _still_ think it's a good idea to keep the component to be added when we send someone to Admin Prison, like I said in the PR body. It's an _admin punishment_, and it is catastrophic when code is introduced that allows subversion of this via in-game processes. I don't see much of any downside to definitely tying it to the target we apply the component to, as it definitely decreases the likelihood of mistakes that allow these kinds of holes to form.

If you 100% disagree and think that I am full of bullshit, I could just keep it as a `do_teleport`. I'm not attached to the work in any way, but I sincerely think that these two principles shouldn't be ignored.

**Comment by exymian** [comment]
"Even Nar'Sie can't help this cultist at this time !"

**Comment by moocowswag** [comment]
> The second is to ensure that the Summon Cultist rune respects `TRAIT_NO_TELEPORT`, so let's do that. 

Will this mean you cant use summon cultist to pull teammates out of luxury shuttle during the occassional shuttle fights where cultists want to grab and convert more people before summoning.

**Comment by Mothblocks** [comment]
To be fair "catastrophic" is overblowing it, admins using the prison are 99.9% of the time watching the guy and are just trying to get them to reply to a damn ticket without killing people

**Comment by Mothblocks** [comment]
That being said I can't tell what your component is actually doing. I think you would have a case if it made sure you would stay in the prison, even when the code failed (and almost certainly runtime/log when it happens so we can replace something with do_teleport), but when area changes, it removes the component, and thus it requires you change the code of summon cultist *anyway*. What's the point of all this?

File: code/datums/components/imprisoned.dm [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+/// Check to see if our area changes at all. If it changes, remove all the no_teleport traits and delete the component.
+/datum/component/imprisoned/proc/verify_imprisonment()
+	var/prisoner_area = get_area(parent)
+	if(prisoner_area == prison_area)
+		return
+	// they're no longer in the prison area, let's automatically remove that trait so they can engage in normal play
+	REMOVE_TRAIT(parent, TRAIT_NO_TELEPORT, ADMIN_TRAIT)
+	qdel(src)
+
+/datum/component/itembound/Destroy(force, silent)
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/datums/components/imprisoned.dm:25]: relative path supporters are tired of winning

[the remark this answers is not in the document]
Comment (san7890) [reply code/datums/components/imprisoned.dm:25]: I copypasted a bit too hard

[the remark this answers is not in the document]
Comment (msgerbs) [reply code/datums/components/imprisoned.dm:25]: Don't copy that floppy!

[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/datums/components/imprisoned.dm:25]: We plan on linting this sandwiched pattern in the future:

```
/path1/a()

/path2/b()

/path1/c()
```

...but would not have caught this.

**Comment by san7890** [comment]
I still think that there's going to be some edge case in the not-so-distant future that adding the trait will solve but w/e

**Comment by optimumtact** [comment]
permanently banned for using [s] for no reason

**Comment by Mothblocks** [comment]
the [s] stands for sex

**Comment by san7890** [comment]
big big chungu[s]

**Pull request merged**